File: Problems/0381-insert-delete-getrandom-o1-duplicates-allowed.py
import random
from collections import defaultdict

class RandomizedCollection(object):

    def __init__(self):
        self.arr = []
        self.pos = defaultdict(set)

    def insert(self, val):
        present = val in self.pos

        self.arr.append(val)
        self.pos[val].add(len(self.arr) - 1)

        return not present

    def remove(self, val):
        if val not in self.pos:
            return False

        idx = self.pos[val].pop()
        last = self.arr[-1]

        if idx != len(self.arr) - 1:
            self.arr[idx] = last
            self.pos[last].remove(len(self.arr) - 1)
            self.pos[last].add(idx)

        self.arr.pop()

        if not self.pos[val]:
            del self.pos[val]

        return True

    def getRandom(self):
        return random.choice(self.arr)

File: Problems/0381-insert-delete-getrandom-o1-duplicates-allowed.cpp
#include <bits/stdc++.h>
using namespace std;

// LeetCode solution starts from here

class RandomizedCollection {
public:
    vector<int> arr;
    unordered_map<int, unordered_set<int>> pos;

    RandomizedCollection() {}

    bool insert(int val) {
        bool present = pos.count(val);

        arr.push_back(val);
        pos[val].insert(arr.size() - 1);

        return !present;
    }

    bool remove(int val) {
        if (!pos.count(val))
            return false;

        int idx = *pos[val].begin();
        pos[val].erase(idx);

        int last = arr.back();

        if (idx != (int)arr.size() - 1) {
            arr[idx] = last;

            pos[last].erase(arr.size() - 1);
            pos[last].insert(idx);
        }

        arr.pop_back();

        if (pos[val].empty())
            pos.erase(val);

        return true;
    }

    int getRandom() {
        return arr[rand() % arr.size()];
    }
};
